src/Workspaces/CSharp/Portable/CodeGeneration/ParameterGenerator.cs

```diff
@@ -140,14 +141,34 @@ internal static ParameterSyntax GetParameter(IParameterSymbol p, CodeGenerationO
         }
 
         private static SyntaxList<AttributeListSyntax> GenerateAttributes(
-            IParameterSymbol parameter, bool isExplicit, CodeGenerationOptions options)
+            IParameterSymbol parameter, bool isExplicit, CodeGenerationOptions options, ref NullableAnnotation nullableAnnotation)
         {
             if (isExplicit)
             {
                 return default;
             }
 
             var attributes = parameter.GetAttributes();
+
+            if (parameter.Type.IsReferenceType)
+            {
+                var newAttributes = attributes.RemoveAll(IsAllowNullAttribute);
+                if (newAttributes.Length != attributes.Length)
+                {
+                    attributes = newAttributes;
+                    if (nullableAnnotation == NullableAnnotation.NotAnnotated)
+                    {
+                        nullableAnnotation = NullableAnnotation.Annotated;
+                    }
+                }
+            }
+            else if (parameter.Type.IsValueType)
+            {
+                attributes = parameter.NullableAnnotation == NullableAnnotation.Annotated
+                    ? attributes.RemoveAll(IsAllowNullAttribute)
+                    : attributes.RemoveAll(IsNullableFlowAnalysisAttribute);
+            }
```

Comment:
1. i don't like all the duplication between methods/properties.
2. i'm not really understanding why the logic is different.  we need to document why the logic is different.

Reply:
Did you mean between methods and parameters?

Reply:
Yes.

Reply:
The current compiler rules don't allow to strip the parameter annotations: https://github.com/dotnet/roslyn/pull/38203#discussion_r364251608, https://github.com/dotnet/roslyn/pull/38203#discussion_r364254604

Reply:
All this information should be documented in the code. Including why each case differs from the other